### DISCUSSION TITLE
Local update in the component that uses store causes an extra update from store

### DISCUSSION THREAD

**Discussion by mahmutyilmazcs** [pmndrs/valtio#638] [Q&A] [answered]
## Summary
Component receives an extra update from the store when its local state changed even though the field it uses from the store never changed.

This is observed in valtio + react usage case and reproducible for both react v16-17 and react v18. With one difference: React 18 optimizes the double update and blocks the one(I guess) 

When following code block is used in the component, you see double commit and double render logs for react v.16.13(also applicable for v17) case and single commit and double render logs for react v18.2 case
```
useEffect(() => {
    console.log("commit");
  });
console.log("render");
```

## Link to reproduction

Please open console and observe logs and profile the app to see an extra update store triggers for both cases.

[react v16.13 + valtio v1.9.0](https://codesandbox.io/s/double-commit-double-render-reactv16-8-tc00ki)

[react v18.2.0 + valtio v1.9.0](https://codesandbox.io/s/double-render-single-commit-reactv18-vw1t30)

Note that both examples above are **NOT** in `Strict Mode`.

**Comment by dai-shi** [comment, marked as answer]
This is expected behavior. If it renders from parent or state, in this case `useState`, we need to run the render and commit, because `bool` in this case may change the result.

The difference between react 18 and pre-18 is also expected. It's because react 18 comes with automatic batching. To get the same behavior in pre-18, one would need to use `unstable_batchedUpdates`.

**Reply by mahmutyilmazcs** [reply]
I was thinking about auto batching feature for React 18 vs pre-18 difference as well. 

But, I don’t quite understand why local state, bool in this case, may change the result. An example would be great.

Thanks.

**Reply by dai-shi** [reply]
For example:
```diff
-   return <div>{"time[never changes]: " + snapshot.time}</div>;
+   return <div>{"time[never changes]: " + snapshot.time * (bool ? 1 : -1)}</div>;
```
will require re-render.

**Reply by mahmutyilmazcs** [reply]
Ok, 'bool' should definitely change the result in your example but changing 'bool'(useState) already does the job for recalculation/re-rendering, right? What's the reason for the second render?

Of course you must have checked the profiler output already but I just want to make sure that I explain myself correctly. Here's the profiler output of the app. Here for every update caused by hook 11, there is an update caused by hook 4 consecutively at almost the same time. Hook 11 is useState('bool') and hook 11 is the useState from valtio store.

<img width="1460" alt="image" src="https://user-images.githubusercontent.com/15721059/215393997-e7a8551b-6500-4e5e-858e-a392c2c8ee6c.png">

<img width="1454" alt="image" src="https://user-images.githubusercontent.com/15721059/215393872-1b5cbb44-2b29-4742-b653-8ebcb31b3522.png">

![image](https://user-images.githubusercontent.com/15721059/215394944-3e675a15-b039-4ef5-817b-405cf793837b.png)

**Reply by dai-shi** [reply]
I think it's still expected, not sure if it's ideal or can be optimized.
setBool triggers re-render, and after the first re-render, `useSnapshot` knows the state is actually changed, but it doesn't know which part of snapshot is used, so triggers re-render to get tracking info, which becomes the second re-render. 

Basically, it works like this:
https://codesandbox.io/s/condescending-carson-u0v3y5?file=/src/App.js

It's with useState, but useSyncExternalStore would be similar.

**Reply by mahmutyilmazcs** [reply]
> but it doesn't know which part of snapshot is used

Doesn't it actually know that only 'time' field is used from the store and only 'count' field is changed. I thought it collects this info from the initial render(mount) of the components. Does it have to check which fields are used for every render?

**Reply by dai-shi** [reply]
No, when `bool` is changed, it may use `count` field (for the first time).

**Reply by mahmutyilmazcs** [reply]
Of course it couldn't know when there is previously unexecuted code blocks. I see...

If I want to block this double render. I have to make 'time' field an object instead of a primitive in the root and use it to generate a snapshot instead of the whole state so snapshot is not changed when 'count' field changes, right?
Is that correct?

```
  // make time field object
  const state = proxy({ count: 0, timeObject: { value: 0 } });

  // create snapshot from the 'timeObject' not from whole 'state' object so that changes in the 'state' (when 'count' field changes) won't trigger an update.

  // use NOT like this
  const snapshot = useSnapshot(state);
  // use like this
  const timeSnapshot = useSnapshot(state.timeObject);

  // use NOT like this
  return <div>{"time[never changes]: " + snapshot.time}</div>
  // use like this
  return <div>{"time[never changes]: " + timeSnapshot.value}</div>
```

**Reply by dai-shi** [reply]
Yeah, or put `bool` in valtio state and eliminate useState. The whole problem is having two different states.

I generally like your approach. It's called scoped subscription.

**Reply by mahmutyilmazcs** [reply]
> Yeah, or put bool in valtio state and eliminate useState. The whole problem is having two different states.

This resolves all problems :) But is not doable. Because, I use a library and it triggers an update inside. Then my component again renders itself because it uses valtio store.

However, let's talk about my approach in the previous answer. In this case every field must be in its own object. This is because, when we add a timezone info to 'time' like this `timeObject: {value: 1, timeZone: 3}`, then, when timezone changes, the component that only uses time.value would get double render again.  So, we have to change it to be `timeObject: {value: {value: 1}, timeZone: {value: 3}}` and subscribe whatever portion of the timeObject you need but it's not maintainable at all.



Morever, I said 

> Of course it couldn't know when there is previously unexecuted code blocks. I see...

in the previous answer but when previously unexecuted code block gets executed for the first time, doesn't it already get the latest value from the store(because it is reading for the first time) and while doing so, valtio can register its listener for that field for further updates?

**Reply by greenwish84** [reply]
Hmm, what an interesting problem😃 

@dai-shi First of all, thanks for implementing this great library that extremely simplifies the global state management 👍 

If I understand correctly, when 2 components (no parent-child relation) have the same `snapshot` call (`useSnapshot(state)`), although each uses separate fields of the state object (1st one "time", 2nd one "count"), any update to a field (ex: `time` field in the 1st component) will trigger a render for the other one (2nd component that only uses `count` field) as well?

If this is the expected behaviour, I think it should be mentioned "specifically" in the documents this way. Also, this may be considered as a little bit of a pitfall for the default valtio usage. I expected that the library would trigger a rerender for a component only for the updates to the fields it use.

As mentioned by @mahmutyilmazcs, object based field subscription (scoped subscription) may not be desirable for both usage and maintainability. I hope for a solution to this issue for the sake of this library's great reputation 😃

**Reply by dai-shi** [reply]
> If I want to block this double render.

But, you don't need to. If double render causes real performance issue, you want to memoize heavy computation.

> Morever, I said

Again, there are two different states. React returns the old state value unless you explicitly update it. In our case, count will be zero. Play with the example in https://github.com/pmndrs/valtio/discussions/638#discussioncomment-4815011. This behavior has nothing to do with valtio. React works like that.

**Reply by mahmutyilmazcs** [reply]
I see why it can't register listener when first read happens, because that first read can be made from one of previous snapshot objects that valtio no longer keeps track of.  I got it.

> > Yeah, or put bool in valtio state and eliminate useState. The whole problem is having two different states.
> 
> This resolves all problems :) But is not doable. Because, I use a library and it triggers an update inside. Then my component again renders itself because it uses valtio store.
> 
> However, let's talk about my approach in the previous answer. In this case every field must be in its own object. This is because, when we add a timezone info to 'time' like this `timeObject: {value: 1, timeZone: 3}`, then, when timezone changes, the component that only uses time.value would get double render again. So, we have to change it to be `timeObject: {value: {value: 1}, timeZone: {value: 3}}` and subscribe whatever portion of the timeObject you need but it's not maintainable at all.
> 
> Morever, I said
> 
> > Of course it couldn't know when there is previously unexecuted code blocks. I see...
> 
> in the previous answer but when previously unexecuted code block gets executed for the first time, doesn't it already get the latest value from the store(because it is reading for the first time) and while doing so, valtio can register its listener for that field for further updates?

**Reply by dai-shi** [reply]
>  when 2 components (no parent-child relation) have the same snapshot call (useSnapshot(state)), although each uses separate fields of the state object (1st one "time", 2nd one "count"), any update to a field (ex: time field in the 1st component) will trigger a render for the other one (2nd component that only uses count field) as well?

No. This only applies when it's rendered by other state.

btw, the said behavior happens in some of my other libs, for example [use-context-selector](https://github.com/dai-shi/use-context-selector) and [use-valtio](https://github.com/dai-shi/use-valtio).
It's still intentional. While it can be improved, it's simple implementation with useReducer.
useReducer works interestingly in React 18: https://twitter.com/dai_shi/status/1534170089981100032

The rationale is double-render-with-single-commit is almost by design, and it's not a problem. The right solution is useMemo.
